Changes to allow the export of functions with no user input.

cptspacemanspiff opened the pull request:
### Summary
This is a set of changes that I have been using to get executorch working with Multiple Entrypoints with shared state.
This is related to: #8030 

### Issue & Changes
Currently trying to export a function:

```python
def init():
  self.shared_state.fill_(0)
```

Will fail, because in to_edge in exir in the program.py the lifted constants are assumed to be placed before the first user input. This causes weird behavior when there are no user inputs. I changed it to place the lifted constant after the last buffer, if there are no user inputs to place before. However, this relies on my understanding of the implicit layout of the graph inputs.

At the same time later in the after the memory planning phase of to_executorch, it validates that memory planning was correct based on whether the `graph_input_allocated` flag is set, this only applies to user inputs, of which we have none, so it errors out. I added a check to bypass this error if there are no user inputs, but I honestly do not understand enough of the validation check to know if that is appropriate.

### Comments

In the current executorch with no support for shared state, this case does not make sense, but #8030 is my attempt at adding that capability. and having initialization methods that init the buffers from constants is useful, especially since their initail state is undefined.

Currently this is not ready, I have no tests/ and have my random notes in the commit as comments, and other than validating that it worked as I was working on the shared state export have not done anything in depth....

But I kind-of want feedback if my solution seems correct, or if I am missing something. Particularly regarding my understanding of placeholder ordering and signature logic, and whether bypassing the graph_input_allocated validation is appropriate.

Out of all the changes I had to make for shared state, this is the one I am least sure about.

 


cc @JacobSzwejbka @angelayi

cptspacemanspiff committed:
WIP need to clean up functions.

JacobSzwejbka commented:
@ydwu4 @angelayi can you guys take a look at this one

JacobSzwejbka commented:
Hmm I thought we already supported functions with no inputs, but I guess before we only tested with constant returns. Now theres no user return either. 

cc @tarun292  who I think worked on the no inputs thing before

JacobSzwejbka commented on exir/memory_planning.py:
```diff
@@ -236,7 +236,10 @@ def verify_graph_input_output(self) -> None:
+        # only check if inputs are allocated if there are user inputs:
+        user_inputs_exist = len(list(filter(lambda input: input.kind == InputKind.USER_INPUT, self.graph_signature.input_specs))) > 0
+
+        if "placeholder" in check_list and user_inputs_exist:
```
this change looks fine. On our side we should probably take a pass over all the memory planning infra. A lot of it was written before export graphs were functionalized so the concept of non user inputs didnt make exist.

angelayi commented on exir/program/_program.py:
```diff
@@ -306,6 +335,9 @@ def lift_constant_tensor_pass(ep):
             new_input_specs.extend(lifted_constants)
```
And if we just insert all the constants at the beginning, we can just directly prepend to input_specs.

cptspacemanspiff replied:
done, and removed all the logic I had previously, now:

No user inputs -> inserting before `None` -> inserts at the beginning.

angelayi commented on exir/program/_program.py:
```diff
@@ -306,6 +335,9 @@ def lift_constant_tensor_pass(ep):
             new_input_specs.extend(lifted_constants)
             lifted_constants.clear()
         new_input_specs.append(s)
```
Can you add a test case?

cptspacemanspiff replied:
I added a test case, shifted to the simpler implementation of just prepending the signature since the order of the non-user inputs does not matter.

cptspacemanspiff replied:
I have also validated that it works with my export process as well.

cptspacemanspiff committed:
Added unit test.

Simplified code, In the case of no user input lifted constants are placed before 'none' node, aka the begining. We update the graph node to place these at the beginning as well.

fixed formatting.

angelayi commented on exir/tests/test_passes.py:
```diff
@@ -1057,6 +1057,35 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+            "b_a"  # followed by the buffer input.
+        ).run(ep.graph_module.code)
+        # the graph signature should also be the same:
+        assert ep.graph_signature.input_specs[0].arg.name == "_lifted_tensor_constant1"
```
```suggestion
        self.assertEquals(ep.graph_signature.input_specs[0].arg.name, "_lifted_tensor_constant1")
```

angelayi commented on exir/tests/test_passes.py:
```diff
@@ -1057,6 +1057,35 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+
+        executorch_program = edge.to_executorch()
+        # # the graph signature should also be the same:
+        # executorch_program.graph_signature.input_specs[0].arg.name == "_lifted_tensor_constant1"
+        # executorch_program.graph_signature.input_specs[1].arg.name == "b_a"
```
why are these commented out?

cptspacemanspiff replied:
I accidentally left those in, when I was initially writing the test/debugging and validating changes I just copy/pasted those lines so I had something to drop a debugger breakpoint on.

I removed those, and left in the to_executorch call, this is there to ensure that the validation steps that were previously failing are correctly bypassed. (Does not throw an exception.)

cptspacemanspiff committed:
Cleaned up test.

cptspacemanspiff commented:
@angelayi

angelayi reviewed (approved):
thanks!!

cptspacemanspiff committed:
Merge branch 'main' into fix-lifted-constants-with-no-user-input

JacobSzwejbka commented:
Im rerunning CI and then assuming it passes Ill merge it

shoumikhin commented:
@cptspacemanspiff any updates? Do we still want this change?
cc @JacobSzwejbka
